src/libsync/clientsideencryption.cpp

```diff
@@ -1315,24 +1438,145 @@ void ClientSideEncryption::sendSignRequestCSR(const AccountPtr &account, PKey ke
                     qCInfo(lcCse()) << ERR_lib_error_string(lastError);
                     lastError = ERR_get_error();
                 }
-                forgetSensitiveData(account);
+                failedToInitialize(account);
                 return;
             }
             qCInfo(lcCse()) << "received a valid certificate";
             fetchAndValidatePublicKeyFromServer(account);
+        } else {
+            qCInfo(lcCse()) << retCode;
+            failedToInitialize(account);
+            return;
+        }
+    });
+    job->start();
+}
+
+void ClientSideEncryption::writeKeyPair(AccountPtr account,
+                                        PKey keyPair,
+                                        QByteArray output)
+{
+    const auto privateKeyKeychainId = AbstractCredentials::keychainKey(
+        account->url().toString(),
+        account->credentials()->user() + e2e_private,
+        account->id()
+        );
+
+    const auto publicKeyKeychainId = AbstractCredentials::keychainKey(
+        account->url().toString(),
+        account->credentials()->user() + e2e_public,
+        account->id()
+        );
+
+    Bio privateKey;
+    if (PEM_write_bio_PrivateKey(privateKey, keyPair, nullptr, nullptr, 0, nullptr, nullptr) <= 0) {
+        qCInfo(lcCse()) << "Could not read private key from bio.";
+        failedToInitialize(account);
+        return;
+    }
+    const auto bytearrayPrivateKey = BIO2ByteArray(privateKey);
+
+    const auto privateKeyJob = new WritePasswordJob(Theme::instance()->appName());
+    privateKeyJob->setInsecureFallback(false);
+    privateKeyJob->setKey(privateKeyKeychainId);
+    privateKeyJob->setBinaryData(bytearrayPrivateKey);
+    connect(privateKeyJob, &WritePasswordJob::finished, [keyPair = std::move(keyPair), publicKeyKeychainId, account, output, this] (Job *incoming) mutable {
+        if (incoming->error() != Error::NoError) {
+            failedToInitialize(account);
+            return;
+        }
+
+        qCInfo(lcCse()) << "Private key stored in keychain";
```

Comment:
@mgallien Do we need these always in logs? Should not we have them as `qCDebug` based on our recent problems with too many logs?

Reply:
no
I cleaned up lots of logs